### DISCUSSION TITLE
outHumAbs = -10,000 instead of None

### DISCUSSION THREAD

**Discussion by Brezentrager** [roe-dl/weewx-GTS#23] [Q&A] [answered]
Hi,
I've included outHumAbs on my website. 

For timestamps with missing temperature and humidity values, outHumAbs results in a value of -10,000 (g/m3).

<img width="525" alt="Bildschirmfoto 2023-03-11 um 20 50 48" src="https://user-images.githubusercontent.com/8404725/224509041-c085f93d-d8de-4949-b7d2-12f56d10c5f5.png">

Am I doing something wrong here?

Thanks!
Manfred

**Comment by roe-dl** [comment, marked as answer]
I am not sure what is happening here. I expected that missing values throw an exception that is catched and sets the result to `None`. 

What is missing in those cases, temperature or humidity or both?

**Reply by Brezentrager** [reply]
Both values are missing.
In the chart, the small dip at 3pm is cased by missing values due to a weewx-restart. The values after 8pm are future points in time.

Yes, reading through the code I also expected to see a  None instead of a negative value.

**Reply by roe-dl** [reply]
Could you include `loginf("%s %s %s" % (temp_C,hum,svp))` into the code and restart WeeWX? 

I guess you will get lots of logging messages out of that. So this can be a short test only.

**Reply by Brezentrager** [reply]
I've put it here (line 529):
<img width="717" alt="Bildschirmfoto 2023-03-11 um 22 39 18" src="https://user-images.githubusercontent.com/8404725/224512641-d1c11a36-af5a-41c7-9cf7-82eb3cae9a40.png">

The resulting values are all okay. No '-10,000' values are printed.

Would there be a better line to put the loginf statement?

**Reply by Brezentrager** [reply]
I've just checked the return values: There are only meaningful values and 'None'. No -10,000 gets returned.

Still, in the json for the Belchertown charts, there are -10,000 values instead.

**Reply by roe-dl** [reply]
Yes, that's the right place to put the statement. Did you find messages where `temp_C` and/or `hum` were `None` or had unreasonable values? 

Is the `outHumAbs` value included in the MQTT output? If so, do you see -10 or `None` there?

You can start WeeWX from command line and see all the packets on screen.

**Reply by Brezentrager** [reply]
Only reasonable values in the logs for svp. (And some `None` when logging the return value of the function)
There were no `None` values for temp_C nor hum.

Yes, it's included in MQTT.
What do you mean with "starting WeeWX from the command line to see the packages"?

**Reply by roe-dl** [reply]
> What do you mean with "starting WeeWX from the command line to see the packages"?

See [WeeWX user's manual](https://weewx.com/docs/usersguide.htm#Running_directly)

**Reply by Brezentrager** [reply]
There is still no indication that the GTS extension would ever produce a value of -10,000.

But now it gets really bizarre:
Yesterday this erroneous value was displayed for the time 3pm (when the temperature and humidity were missing). Just now, however, I saw that on the page with yesterday's charts, the values for `outHumAbs` are now correctly shown as `None` for 3 pm. See here: https://www.wetter-zorneding.de/graphs/?graph=yesterday

However, a similar issue exists for 14 February, as can be seen in the monthly summary. In that time period, `humidity` was missing: https://www.wetter-zorneding.de/graphs/?graph=lastmonth

**Reply by roe-dl** [reply]
Interesting. Could there be some memory problem with your computer?

If I look at `https://www.wetter-zorneding.de/json/lastmonth.json` I see the values there:
```
                   [
                        1676322000000.0,
                        4403695.0
                    ],
                    [
                        1676325600000.0,
                        4268478.0
                    ],
                    [
                        1676329200000.0,
                        -10000000000.0
                    ],
                    [
                        1676332800000.0,
                        -10000000000.0
                    ],
                    [
                        1676336400000.0,
                        -10000000000.0
                    ],
                    [
                        1676340000000.0,
                        4050694.0
                    ],
                    [
                        1676343600000.0,
                        4069515.0
                    ],
```

May be, you want to have a look at `user/belchertown.py` by including some debugging statements there.

**Reply by Brezentrager** [reply]
Yes, those are the values for 14 February. No clue where they come from.

I already tried to put some debugging statements in `belchertown.py`, but wasn't able to locate the best place to put them.
I also tried to deploy a "dirty" fix for the problem, by checking values before they get written to the json files. But also here I didn't find, where exactly I have to do that.

I'll try once again tonight.

Thanks!

**Reply by roe-dl** [reply]
The JSON files are created in class `HighchartsJsonGenerator`. There the function `get_observation_data()` fetches the data from the database. Within this function you find `get_series()` called. A little bit later there is a statement about `special_target_unit`. I would try to include the debugging statements after that statement. In my case that is line 3729, but I use an older version of `belchertown.py`.

**Reply by Brezentrager** [reply]
I've now included the dirty fix for the problem in `Belchertown.py`:

<img width="971" alt="Bildschirmfoto 2023-03-12 um 22 31 33" src="https://user-images.githubusercontent.com/8404725/224574920-90efbe25-2063-4fc2-a7e2-11bca7d4660d.png">

All negative values for `outHumAbs` are replaced by `None`.

Thanks for explaining where to best put the fix and for all your effort! Highly appreciated!
